PAM: various improvements

valpackett opened the pull request:
Continuing from #14789… @felixdoerre

### Motivation and Context

- *PAM: add 'recursive_homes' flag to use with 'prop_mountpoint'* – I really need to not have the limitation of having all homes under one directory, that limitation doesn't really make sense with `prop_mountpoint` IMO.
- *PAM: support password changes even when not mounted* – had this patch around for a while, not the most important for me personally anymore but thought it might still be useful for others, because it makes sense that the `passwd` command shouldn't require a user to be logged in for an admin to change their password.
- *PAM: add 'forceunmount' flag* – honestly mostly because the unmount was always failing after exiting a simple FreeBSD's `login` session (though e.g. not a `su -l` one) heh.
- *PAM: do not fail to mount if the key's already loaded* – well, motivated entirely by the mount failing after such an unmount failure, but one could imagine a variety of other situations where this could happen.
- *PAM: add 'uid_min' and 'uid_max' options for changing the uid range* – the inability to have root have a pam_zfs_key'ed home due to the `uid >= 1000` check was bothering me, so I've decided to convert that check into a configurable one.

### Description

A variety of PAM module changes. Submitting as one PR to avoid having a bunch of barely-different CI runs, and they're not that complex anyway. If only some commits are acceptable, I can remove the others. Bikeshedding on option names also definitely welcome :)

### How Has This Been Tested?

~~"Works for me"…~~ Actually now added tests for password change when unmounted and for recursive search.

### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

behlendorf reviewed (approved):
These changes all look reasonable to me.

behlendorf commented:
However, one of the existing pam tests [is reliably failing](http://build.zfsonlinux.org/builders/CentOS%207%20x86_64%20%28TEST%29/builds/24287/steps/shell_4/logs/summary) after this change.  That will need to be sorted out before this can be merged.  Additional pam related test cases for the new functionality here would be a good ideas as well.

```
Tests with results other than PASS that are unexpected:
    FAIL pam/pam_short_password (expected PASS)
````

felixdoerre commented:
> PAM: add 'everywhere' flag to use with 'prop_mountpoint' – this one is the most important to my use case, as I want to support portable home directories, i.e. mounting ZFS from an external USB drives and logging in as a user with a home directory there – obviously a fixed single parent dataset is not sufficient for this. And in general it's kinda annoying to specify one ;) if we're using prop_mountpoint search anyway.

I am not sure I understand the usecase/security model here. For this to work you would need something to automatically import your zpool on plug-in, right?

So if someone is at your computer they just need some removable drive with a dataset that has your home directory as mountpoint and anything set as passphrase, to be able to login, right? Paired with sudo permissions on that particular computer feels dangerous to me.

For local, non-physical attackers this might not be a problem, as currently only "root" is allowed to create new datasets and/or import pools, correct?

valpackett commented:
> For this to work you would need something to automatically import your zpool on plug-in, right?

No, only someone with existing root permissions would import the zpool… Yeah anyway there's no fully thought out security model for *that* yet, I shouldn't even have mentioned that use case :) Actually… this should be done a little bit differently, there should be a `recursive` flag and the ability to set the homes root to "everything" as orthogonal options, so it would be possible to e.g. only allow everything under one pool.

---

Wrangling the test stuff right now…

ryao commented on contrib/pam_zfs_key/pam_zfs_key.c:
```diff
@@ -721,49 +755,61 @@ pam_sm_chauthtok(pam_handle_t *pamh, int flags,
-		free(dataset);
-		pam_zfs_free();
-		if (! key_loaded) {
+		if (decrypt_mount(pamh, dataset, old_token->value, B_TRUE) == -1) {
```
cstyle is also failing in two places due to lines being too long. Please wrap these.

valpackett replied:
Yeah I saw that… still working on more tests btw.

…is it just me or does checkstyle count a tab as 4/8 characters instead of 1? :/

behlendorf replied:
It's not you, it counts a tab as 8.

valpackett commented:
Refactored the recursive vs everywhere thing into two orthogonal things; added tests for recursive homes and password change when unmounted.

Not sure what's up with the CentOS fails but also… where did the FreeBSD buildbot go? :( Testing PAM on FreeBSD is also something I wanted to enable.

behlendorf commented:
The FreeBSD builder were offline for a bit if you can rebase again and force update the PR they should run.

valpackett commented:
Yay, the pam tests passed on the server 🎉

behlendorf commented:
@felixdoerre would you mind taking another look at this.  If you don't have any concerns I'd like to merge it.

felixdoerre reviewed (approved):
The changes look good to me. The only thing I am wondering is, if we should start a man page for the pam options. Especially, as homes now supports a special value (`homes=*`).
